elastic/kibana
# [SLO] [Alert detail page] render log rate analysis for custom kql

Fixes https://github.com/elastic/kibana/issues/178446

This PR adds the following AI Ops features to the Burn rate alert detail page:

- Log Rate Analysis
- ML features / AI Assistant

https://github.com/elastic/kibana/assets/2852703/bf00c636-10ab-4cd8-859b-4a8f0471bf15

## Implementation details
Log rate analysis is useful when there is a significant dip/spike in number of logs. It finds significant differences between the logs in baseline (before dip/spike) and in deviation (after dip/spike). For the Log Rate Analysis to work on the Alert details of an SLO burn rate rule we had to specify a few params, here are the most important ones:
- dataView
- esSearchQuery
- timeRange 
- initialAnalysisStart

For the esQuery here's what we used

```
const finalQuery = {
    bool: {
      filter: [customTotalFilter, customFilters, customGroupByFilters],
      must_not: customGoodFilter,
    },
  };
```

We add the `total query`, any `optional filters` and any `group by filters` to the `filter` clause of the ES query, where as we  add the good filter in the `must_not` clause.

## 🔬 How to test
- Add the following lines to your `config/kibana.dev.yaml`:
  - `server.basePath: '/kibana'`
  - `server.publicBaseUrl: 'http://localhost:5601/kibana'`
- Start with the following command: `node x-pack/scripts/data_forge.js --events-per-cycle 50 --lookback now-1d --dataset fake_stack --install-kibana-assets --kibana-url http://localhost:5601/kibana --event-template good` to start with some good events
- Wait til the log message says `info Waiting 60000ms`
- Create one SLO:
  - "Admin Console Availability" using the "Custom Query" SLI with the `Admin Console` DataView, set the "Good query" to `http.response.status_code < 500` and the set the "Total query" to `http.response.status_code: *` using a rolling `7d` time window
- You should have 1 burn rate rule created by default
- Stop the first `data_forge.js` command
- Start `node x-pack/scripts/data_forge.js --events-per-cycle 50 --lookback now --dataset fake_stack --install-kibana-assets --kibana-url http://localhost:5601/kibana --event-template bad`
- Go to alert details page of the burn rate rule to verify that the Log Rate Analysis tool detected a spike  

Note: you can skip first step, but in this case you should start kibana with this command instead `yarn start --no-base-path`. This way you don't have to pass `--kibana-url` when running the `data_forge` command

I included some snapshot tests with different SLOs that could be used for verifying the generation of the ES query. Here are a few examples:
- Good query (KQL), no filter, no total query
- Good query (filter), no optional filter, no total query
- Good query (kql) optional filter (KQL), no total query
- Good query (kql), optional filter (filter), no total query
- Good query (kql), optional filter (kql), total query (kql)
- Good query (filter), optional filter(filter), total query(kql)

💀 I will need extra help testing this feature, to make sure that provided params to the `LogRateAnalysisContent` component. Especially `esSearchQuery` and `timeRange` (timeRange still WIP, but I want some early feedback on the lookbackDuration and the intervalFactor) 

## ✔️ Acceptance criteria
- Render the Log Rate Analysis tool only to platinum users and above
- Render AI assistant contextual insights only to platinum users and above
